### PR TITLE
Add an incremental API for running benchmarks

### PR DESCRIPTION
**Pull request by bens**
Allows to get results as they become available and show a progress bar type display, and allows to limit memory use by freeing results when they're received.

### CODE REVIEW

File: zbench.zig [hunk 1]
```diff
@@ -89,6 +89,71 @@ pub const Benchmark = struct {
...
+
+            const ns = try self.b.runFunc(self.benchmarks[0]);
+            if (try r.next(ns)) |_| {
+                const total_benchmarks = self.b.benchmarks.items.len;
+                return Step{ .progress = Progress{
+                    .total_benchmarks = total_benchmarks,
+                    .completed_benchmarks = total_benchmarks - self.benchmarks.len,
+                    .current_name = self.benchmarks[0].name,
+                    .total_runs = 0,
+                    .completed_runs = 0,
```
Comment (bens) [inline zbench.zig:132]: Ah, this was a placeholder, need to figure out how it will work.

File: zbench.zig [hunk 2]
```diff
@@ -89,6 +89,71 @@ pub const Benchmark = struct {
         });
     }
 
+    pub const Iterator = struct {
+        allocator: std.mem.Allocator,
+        b: *const Benchmark,
+        benchmarks: []const Definition,
```
Comment (bens) [inline zbench.zig:95]: This is "benchmarks still to run", so needs a better name.

File: zbench.zig [hunk 3]
```diff
@@ -89,6 +89,71 @@ pub const Benchmark = struct {
...
+        runner: ?Runner,
```
Comment (bens) [inline zbench.zig:96]: I should add an `abort()` method to clean up the Runner in an `errdefer`.

File: zbench.zig [hunk 4]
```diff
@@ -89,6 +89,96 @@ pub const Benchmark = struct {
...
+    pub const Iterator = struct {
+        allocator: std.mem.Allocator,
+        b: *const Benchmark,
+        remaining: []const Definition,
+        runner: ?Runner,
+
+        /// A summary of progress through benchmarking a function. The
+        /// total_runs field may vary over time as the runner calibrates how
+        /// many runs to perform.
+        pub const Progress = struct {
```
Comment (hendriknielaender) [inline zbench.zig:102]: What do you think about using the std lib for displaying the progress?
https://ziglang.org/documentation/master/std/#std.Progress

Used it in [another project](https://github.com/hendriknielaender/zvm/pull/22/files), came in handy.

Comment (bens) [reply zbench.zig:102]: Nice suggestion, it's pretty easy to use from following the example, I've added it to the sample code.

File: examples/basic.zig [hunk 5]
```diff
@@ -22,26 +10,27 @@ fn afterAllHook() void {
...
         },
     });
-
-    const sysinfo = try bench.getSystemInfo();
-    try std.fmt.format(stdout, "\n{}\n", .{sysinfo});
-
-    const results = try bench.run();
-    defer results.deinit();
-    try results.prettyPrint(stdout, true);
+    try bench.run(stdout);
```
Comment (bens) [inline examples/basic.zig:35]: Any objection to doing the pretty printing and handling std.Progress in the `run()` function? I think it's ok for it to do more for the common case when there's the `iterator()` method which gives lower level access.

Comment (hendriknielaender) [reply examples/basic.zig:18]: No objections, i think thats justified.

File: zbench.zig (hunk 4 above)
Comment (hendriknielaender) [reply zbench.zig:102]: Cool 👍

**Comment by bens** [comment]
I think this one's ready to merge, unless there's anything else.

**Pull request merged**